import sys
import threading
import os
import logging
import datetime
from zoneinfo import ZoneInfo

from PySide6.QtWidgets import QApplication, QWidget, QMessageBox
from PySide6.QtCore import QFile
from PySide6.QtUiTools import QUiLoader
from PySide6.QtGui import QIcon
from PySide6.QtWidgets import QVBoxLayout

from utils import get_online_time
from crawler import crawl
from logging_handler import LogEmitter, QTextBrowserHandler

class MainWindow(QWidget):
    def __init__(self):
        super().__init__()

        # .ui íŒŒì¼ ë¡œë“œ ì¤€ë¹„
        base_path = getattr(sys, '_MEIPASS', os.path.dirname(__file__))
        ui_path   = os.path.join(base_path, 'budongsan.ui')
        ui_file   = QFile(ui_path)
        if not ui_file.open(QFile.ReadOnly):
            raise RuntimeError(f"UI íŒŒì¼ì„ ì—´ ìˆ˜ ì—†ìŠµë‹ˆë‹¤: {ui_path}")

        # QUiLoaderë¡œ ìœ„ì ¯ ìƒì„±
        loader = QUiLoader()
        self.ui = loader.load(ui_file)       # â† ë°˜í™˜ëœ ìœ„ì ¯ì„ self.uiì— ì €ì¥
        ui_file.close()

        # QMainWindow ì¤‘ì•™ ìœ„ì ¯ìœ¼ë¡œ ë°°ì¹˜
        layout = QVBoxLayout(self)
        layout.addWidget(self.ui)
        self.setWindowTitle("ì‚¼ì‚¼ì— íˆ¬ í¬ë¡¤ë§ ìë™í™” í”„ë¡œê·¸ë¨")

        # ë””ìì´ë„ˆì—ì„œ ì„¤ì •í•œ ê¸°ë³¸ í¬ê¸°ë¡œ ì°½ í¬ê¸° ì¡°ì •
        self.resize(self.ui.size())


        # ì°½ ì„¤ì •
        self.setWindowIcon(QIcon("budongsan_icon.png"))
        self.set_expiration_date()

        # ë²„íŠ¼ ì‹œê·¸ë„ ì—°ê²°
        self.ui.keyword_btn.clicked.connect(self.keyword_page_open)
        self.ui.start_btn.clicked.connect(self.start_crawling)
        self.ui.reset_btn.clicked.connect(self.reset_fields)
        self.ui.quit_btn.clicked.connect(self.quit_application)

        # ê¸°íƒ€ ì´ˆê¸°í™”
        self.keywords = []
        self.setup_logging()

    def set_expiration_date(self):
        expiration_date = datetime.datetime(2045, 1, 1, tzinfo=ZoneInfo("Asia/Seoul"))
        online_time = get_online_time()
        today = online_time or datetime.datetime.now(ZoneInfo("Asia/Seoul"))

        if today > expiration_date:
            self.ui.expire_date.setText("í”„ë¡œê·¸ë¨ ì‚¬ìš© ê¸°í•œì´ ë§Œë£Œë˜ì—ˆìŠµë‹ˆë‹¤.")
            self.ui.start_btn.setEnabled(False)
        else:
            remaining_days = (expiration_date - today).days
            self.ui.expire_date.setText(
                f"{expiration_date.strftime('%Y-%m-%d')} (ë‚¨ì€ ì¼ìˆ˜: {remaining_days}ì¼)"
            )

    def setup_logging(self):
        self.log_emitter = LogEmitter()
        self.log_emitter.log_signal.connect(self.update_status)

        handler = QTextBrowserHandler(self.log_emitter)
        handler.setFormatter(logging.Formatter('%(asctime)s - %(levelname)s - %(message)s'))

        logging.getLogger().addHandler(handler)
        logging.getLogger().setLevel(logging.INFO)

    def keyword_page_open(self):
        keyword_input = self.ui.keyword.text().strip()
        if not keyword_input:
            QMessageBox.warning(self, "ì…ë ¥ ì˜¤ë¥˜", "í‚¤ì›Œë“œë¥¼ ì…ë ¥í•´ì£¼ì„¸ìš”.")
            return

        self.keywords = [kw.strip() for kw in keyword_input.split(',') if kw.strip()]
        if not self.keywords:
            QMessageBox.warning(self, "ì…ë ¥ ì˜¤ë¥˜", "ìœ íš¨í•œ í‚¤ì›Œë“œë¥¼ ì…ë ¥í•´ì£¼ì„¸ìš”.")
            return

        QMessageBox.information(self, "í‚¤ì›Œë“œ ì„¤ì •", f"ì„¤ì •ëœ í‚¤ì›Œë“œ: {', '.join(self.keywords)}")

    def start_crawling(self):
        if not self.keywords:
            QMessageBox.warning(self, "ì‹œì‘ ì˜¤ë¥˜", "ë¨¼ì € í‚¤ì›Œë“œë¥¼ ì„¤ì •í•´ì£¼ì„¸ìš”.")
            return

        self.ui.start_btn.setEnabled(False)
        self.ui.keyword_btn.setEnabled(False)
        self.ui.textBrowser.clear()

        threading.Thread(target=self.run_crawling, daemon=True).start()

    def run_crawling(self):
        try:
            logging.info(f"Starting crawl for keywords: {', '.join(self.keywords)}")

            # ì´ë¯¸ì§€ ê¸°ë³¸ í´ë”
            base_image_dir = "images"
            os.makedirs(base_image_dir, exist_ok=True)

            # ì—‘ì…€ ì¶œë ¥ í´ë”
            os.makedirs("output", exist_ok=True)

            # ğŸš€ keyword ë¦¬ìŠ¤íŠ¸ ì „ì²´ë¥¼ í†µì§¸ë¡œ crawl()ì— ì „ë‹¬
            crawl(self.keywords, base_image_dir, "output")

            logging.info("ëª¨ë“  í‚¤ì›Œë“œ í¬ë¡¤ë§ ì™„ë£Œ")
            self.update_status("ëª¨ë“  í‚¤ì›Œë“œ í¬ë¡¤ë§ ì™„ë£Œ")

        except Exception as e:
            logging.error(f"Error crawling: {e}", exc_info=True)
            self.update_status("Error crawling")

        self.enable_buttons()


    def update_status(self, message):
        self.ui.textBrowser.append(message)

    def enable_buttons(self):
        self.ui.start_btn.setEnabled(True)
        self.ui.keyword_btn.setEnabled(True)

    def reset_fields(self):
        self.ui.keyword.setText('')
        self.ui.textBrowser.clear()
        self.keywords = []

    def quit_application(self):
        sys.exit()

if __name__ == "__main__":
    app = QApplication(sys.argv)
    app.setWindowIcon(QIcon("budongsan_icon.png"))
    window = MainWindow()
    window.show()
    sys.exit(app.exec())
